File: Algorithms/Flow/Algorithms/Part_1_Simple_Algorithms/2.Generating subsets/Generating_Subsets.cpp
//
//  Generating_Subsets.cpp
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/5/24.
//


#include <iostream>
#include <vector>

void generateSubsets(
					 int index,
					 std::vector<int>& subset,
					 const std::vector<int>& nums
					 )
{
	// Print the current subset
	std::cout << "{ ";
	for (int num : subset)
	{
		std::cout << num << " ";
	}
	std::cout << "}" << std::endl;

	// Generate subsets starting from the current index
	for (int i = index; i < nums.size(); ++i)
	{
		// Include nums[i] in the current subset
		subset.push_back(nums[i]);
		// Recur for the next elements
		generateSubsets(i + 1, subset, nums);
		// Backtrack to explore other possibilities
		subset.pop_back();
	}
}

int main() {
	std::vector<int> subset; // To hold the current subset
	std::vector<int> nums = {1, 2, 3}; // Input set

	generateSubsets(0, subset, nums); // Start generating subsets

	return 0;
}

File: Algorithms/Flow/Algorithms/Part_1_Simple_Algorithms/2.Generating subsets/GeneratingSubsets.swift
//
//  GeneratingSubsets.swift
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/5/24.
//

import Foundation

func generateSubsets(index: Int, subset: inout [Int], nums: [Int]) {
	// Print the current subset
	print("{ \(subset.map(String.init).joined(separator: " ")) }")
	
	// Generate subsets starting from the current index
	for i in index..<nums.count {
		// Include nums[i] in the current subset
		subset.append(nums[i])
		// Recur for the next elements
		generateSubsets(index: i + 1, subset: &subset, nums: nums)
		// Backtrack to explore other possibilities
		subset.removeLast()
	}
}

/*
Example:
 func main() {
	var subset: [Int] = [] // To hold the current subset
	let nums = [1, 2, 3] // Input set

	generateSubsets(index: 0, subset: &subset, nums: nums) // Start generating subsets
}

main()

*/
